external_search.py

```diff
@@ -882,8 +577,748 @@ def create_new_version(cls, search_client, base_index_name, version=None):
+        vocabulary, and remove it from the query string, so it
+        doesn't get reused later.
+        """
+        # If the match was "children" and the query string was
+        # "children's", we want to remove the "'s" as well as
+        # the match. We want to remove everything up to the
+        # next word boundary that's not an apostrophe or a
+        # dash.
+        word_boundary_pattern = r"\b%s[\w'\-]*\b"
+
+        return re.compile(
+            word_boundary_pattern % match.strip(), re.IGNORECASE
+        ).sub("", query_string)
+
+
+class Filter(SearchBase):
+    """A filter for search results.
+
+    This covers every reason you might want to not show a search
+    result that matches the query string -- wrong media, wrong
+    language, not available in the patron's library, etc.
+    """
+
+    @classmethod
+    def from_worklist(cls, _db, worklist, facets):
+        """Create a Filter that finds only works that belong in the given
+        WorkList and EntryPoint.
+
+        :param worklist: A WorkList
+        :param facets: A SearchFacets object.
+        """
+        library = worklist.get_library(_db)
+
+        v = worklist.inherited_value
+        media = v('media')
+        languages = v('languages')
+        fiction = v('fiction')
+        audiences = v('audiences')
+        target_age = v('target_age')
+
+        # For genre IDs and CustomList IDs, we might get a separate
+        # set of restrictions from every item in the WorkList hierarchy.
+        # _All_ restrictions must be met for a work to match the filter.
+        v = worklist.inherited_values
+        genre_id_restrictions = v('genre_ids')
+        customlist_id_restrictions = v('customlist_ids')
+        return cls(
+            library, media, languages, fiction, audiences,
+            target_age, genre_id_restrictions, customlist_id_restrictions,
+            facets
+        )
+
+    def __init__(self, collections=None, media=None, languages=None,
+                 fiction=None, audiences=None, target_age=None,
+                 genre_restriction_sets=None, customlist_restriction_sets=None,
+                 facets=None
+    ):
+
+        if isinstance(collections, Library):
+            # Find all works in this Library's collections.
+            collections = collections.collections
+        self.collection_ids = self._filter_ids(collections)
+
+        self.media = media
+        self.languages = languages
+        self.fiction = fiction
+        self.audiences = audiences
+
+        if target_age:
+            if isinstance(target_age, int):
+                self.target_age = (target_age, target_age)
+            elif isinstance(target_age, tuple) and len(target_age) == 2:
+                self.target_age = target_age
+            else:
+                # It's a SQLAlchemy range object. Convert it to a tuple.
+                #
+                # NOTE: This assumes that the range is inclusive
+                # on both sides.
```

Comment:
Could you use `numeric_range_to_tuple` from model.py to remove this assumption? I recently fixed a bug due to inclusive ranges becoming exclusive on commit, so I wouldn't trust this.

Reply:
Good idea.

Reply:
Wow, I think I uncovered a bug in `numericrange_to_tuple`.